File: 13_integrated_testing/tests/00_interface_sanity_console.rb
# frozen_string_literal: true

# SPDX-License-Identifier: MIT OR Apache-2.0
#
# Copyright (c) 2019-2020 Andre Richter <andre.o.richter@gmail.com>

require 'expect'

TIMEOUT_SECS = 3

# Verify sending and receiving works as expected.
class TxRxHandshake
    def name
        'Transmit and Receive handshake'
    end

    def run(qemu_out, qemu_in)
        qemu_in.write_nonblock('ABC')
        raise('TX/RX test failed') if qemu_out.expect('OK1234', TIMEOUT_SECS).nil?
    end
end

# Check for correct TX statistics implementation. Depends on test 1 being run first.
class TxStatistics
    def name
        'Transmit statistics'
    end

    def run(qemu_out, _qemu_in)
        raise('chars_written reported wrong') if qemu_out.expect('6', TIMEOUT_SECS).nil?
    end
end

# Check for correct RX statistics implementation. Depends on test 1 being run first.
class RxStatistics
    def name
        'Receive statistics'
    end

    def run(qemu_out, _qemu_in)
        raise('chars_read reported wrong') if qemu_out.expect('3', TIMEOUT_SECS).nil?
    end
end

##--------------------------------------------------------------------------------------------------
## Test registration
##--------------------------------------------------------------------------------------------------
def subtest_collection
    [TxRxHandshake.new, TxStatistics.new, RxStatistics.new]
end

File: 13_integrated_testing/tests/00_interface_sanity_console.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// Copyright (c) 2019-2020 Andre Richter <andre.o.richter@gmail.com>

//! Console sanity tests - RX, TX and statistics.

#![feature(format_args_nl)]
#![no_main]
#![no_std]

mod panic_exit_failure;

use libkernel::{bsp, interface::console::*, print};

#[no_mangle]
unsafe fn kernel_init() -> ! {
    bsp::qemu_bring_up_console();

    // Handshake
    assert_eq!(bsp::console().read_char(), 'A');
    assert_eq!(bsp::console().read_char(), 'B');
    assert_eq!(bsp::console().read_char(), 'C');
    print!("OK1234");

    // 6
    print!("{}", bsp::console().chars_written());

    // 3
    print!("{}", bsp::console().chars_read());

    // The QEMU process running this test will be closed by the I/O test harness.
    loop {}
}
